src/download.c

```diff
@@ -109,6 +112,27 @@ struct multi_curl_file {
 	void *data; /* user's data */
 };
 
+/*
+ * Used for wrapping the success callback and tracking return values from threads
+ * and a pointer to the associated return value's file data, to re-schedule on failure
+ */
+struct multithread_data {
+	struct multi_curl_file *file;   /* Curl file information */
+	swupd_curl_success_cb callback; /* Holds original success callback to be wrapped */
+
+	bool retval; /*return value from callback*/
```

Comment:
Can't you move retval to struct multi_curl_file and then you won't need to have this struct?
And instead of mt_handles you can iterate curl_hashmap using HASHMAP_FOREACH to check all files.
Then you can get rid of the success wrapper too